<a href="https://colab.research.google.com/github/ge96lip/Quantum-Computing/blob/main/QC_Shor's_Factorization_Algorithm.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
#!pip install qiskit-ibm-runtime

from qiskit_ibm_runtime import QiskitRuntimeService
service = QiskitRuntimeService.save_account(channel="ibm_quantum", token="86ad51ed8e455dea40b8ac15c7b6359cf38cb674fdb3072f3d2e6211debf389fcfe547c54aad314614600bbe71f3d83572bc7059184a3bb11feae345ec347801", overwrite=True)

from qiskit_ibm_runtime import QiskitRuntimeService
service = QiskitRuntimeService()

from qiskit import QuantumCircuit
bell = QuantumCircuit(2)

bell.h(0)
bell.cx(0, 1)
#bell.measure_all()

print(bell.draw())

backend = service.least_busy(operational=True, simulator=False, min_num_qubits=2)
print(backend.name)

from qiskit import transpile

bell_transpiled = transpile(bell, backend, optimization_level=2)
print(bell_transpiled.draw())

from qiskit_ibm_runtime import SamplerV2 as Sampler
sampler = Sampler(backend)
bell_job = sampler.run([bell_transpiled], shots=1000)

print(bell_job.result()[0].data.meas.get_counts())
from qiskit.visualization import plot_histogram
plot_histogram(bell_job.result()[0].data.meas.get_counts())

     ┌───┐     
q_0: ┤ H ├──■──
     └───┘┌─┴─┐
q_1: ─────┤ X ├
          └───┘
ibm_brisbane
global phase: 3π/4
                                                           
    ancilla_0 -> 0 ────────────────────────────────────────
                                                           
    ancilla_1 -> 1 ────────────────────────────────────────
                                                           
    ancilla_2 -> 2 ────────────────────────────────────────
                                                           
    ancilla_3 -> 3 ────────────────────────────────────────
                                                           
    ancilla_4 -> 4 ────────────────────────────────────────
                                                           
    ancilla_5 -> 5 ────────────────────────────────────────
                                                           
    ancilla_6 -> 6 ────────────────────────────────────────
                                                

/usr/local/lib/python3.10/dist-packages/qiskit_ibm_runtime/utils/validations.py:40: UserWarning: The 0-th circuit has no output classical registers so the result will be empty. Did you mean to add measurement instructions?
  warnings.warn(


KeyboardInterrupt: 

In [4]:
# Loading your IBM Q account
QiskitRuntimeService.save_account(channel="ibm_quantum", token='86ad51ed8e455dea40b8ac15c7b6359cf38cb674fdb3072f3d2e6211debf389fcfe547c54aad314614600bbe71f3d83572bc7059184a3bb11feae345ec347801', overwrite=True)

# Run the circuit on a IBM quantum computer
service = QiskitRuntimeService()
#backend = service.least_busy(operational=True, simulator=False, min_num_qubits=20)
backend = service.backend("ibm_brisbane")
print("Selected backend:", backend.name)

# Create a simple quantum circuit
qc_test = QuantumCircuit(2)
qc_test.h(0)
qc_test.cx(0, 1)
#qc_test.measure_all()
print(qc_test)

# Transpile the circuit for the selected backend
transpiled_qc = transpile(qc_test, backend)
#print(transpiled_qc.draw())

#transpiled_qc = transpile(qc_test, backend)
sampler = Sampler(mode=backend)
job = sampler.run([transpiled_qc], shots=1)



Selected backend: ibm_brisbane
     ┌───┐     
q_0: ┤ H ├──■──
     └───┘┌─┴─┐
q_1: ─────┤ X ├
          └───┘
global phase: 7π/4
                    ┌────────┐ ┌────┐          ┌──────┐┌──────────┐┌────┐»
          q_0 -> 0 ─┤ Rz(-π) ├─┤ √X ├──────────┤1     ├┤ Rz(-π/2) ├┤ √X ├»
                   ┌┴────────┴┐├────┤┌────────┐│  Ecr │├─────────┬┘├────┤»
          q_1 -> 1 ┤ Rz(-π/2) ├┤ √X ├┤ Rz(-π) ├┤0     ├┤ Rz(π/2) ├─┤ √X ├»
                   └──────────┘└────┘└────────┘└──────┘└─────────┘ └────┘»
    ancilla_0 -> 2 ──────────────────────────────────────────────────────»
                                                                         »
    ancilla_1 -> 3 ──────────────────────────────────────────────────────»
                                                                         »
    ancilla_2 -> 4 ──────────────────────────────────────────────────────»
                                                                         »
    ancilla_3 -> 5 ──────────────────────────

/usr/local/lib/python3.10/dist-packages/qiskit_ibm_runtime/utils/validations.py:40: UserWarning: The 0-th circuit has no output classical registers so the result will be empty. Did you mean to add measurement instructions?
  warnings.warn(


# Prerequisites

In [1]:
!pip install qiskit
!pip install pylatexenc
!pip install qiskit_aer
!pip install qiskit-ibm-runtime
!pip install qiskit qiskit-ibm-runtime

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.8/4.8 MB 22.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 119.4/119.4 kB 4.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 10.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.7/49.7 kB 607.1 kB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.7/49.7 MB 8.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 108.5/108.5 kB 2.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 162.6/162.6 kB 4.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for pylatexenc: filename=pylatexenc-2.10-py3-none-any.whl size=136816 sha256=7ae0413c4c1f99078e5b50cb7e4cc951e908dbe029da1e9ce21c1502a396a107
  Stored in directory: /root/.cache/pip/wheels/d3/31/8b/e09b0386afd80cfc556c00408c9aeea5c35c4d484a9c762fd5
Successfully built pylatexenc
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.3/12.3 MB 70.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

In [2]:
from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister, transpile
from qiskit.circuit import Qubit
from qiskit.circuit.library import XGate, CSwapGate, QFT
from qiskit import transpile
from qiskit_aer import AerSimulator
from qiskit_ibm_runtime import QiskitRuntimeService, Sampler
from qiskit.visualization import plot_histogram

import matplotlib.pyplot as plt
import math
import time

# Loading your IBM Q account
QiskitRuntimeService.save_account(channel="ibm_quantum", token='86ad51ed8e455dea40b8ac15c7b6359cf38cb674fdb3072f3d2e6211debf389fcfe547c54aad314614600bbe71f3d83572bc7059184a3bb11feae345ec347801', overwrite=True)


# Shor's Factorization Algorithm

In [3]:
def ShorLogic(N, repeat_period_candidates, coprime):
    print('Repeat period candidates: '+str(repeat_period_candidates)+'\n')
    factor_candidates = []
    for i in range(len(repeat_period_candidates)):
        repeat_period = repeat_period_candidates[i];
        # Given the repeat period, find the actual factors
        ar2 = pow(coprime, repeat_period / 2.0)
        factor1 = int(gcd(N, ar2 - 1))
        factor2 = int(gcd(N, ar2 + 1))
        factor_candidates.append([factor1, factor2])
    return factor_candidates

In [4]:
def gcd(a, b):
    # return the greatest common divisor of a,b
    while b:
        m = a % b
        a = b
        b = m
    return a

In [5]:
def check_result(N, factor_candidates):
    for i in range(len(factor_candidates)):
        factors = factor_candidates[i]
        if factors[0] * factors[1] == N:
            if factors[0] != 1 and factors[1] != 1:
                # Success!
                return factors
    # Failure
    return None

In [6]:
def Shor(N, coprime, classical=False, simulate=True, vis=False, opt_for_vis=False):
    # Quantum part
    if classical:
        # Classical alternative for comparision
        repeat_period = ShorNoQPU(N, coprime)
    else:
        # Quantum part
        repeat_period = ShorQPU(N, coprime, simulate, vis, opt_for_vis)

    # Classical part
    factors = ShorLogic(N, repeat_period, coprime)

    return check_result(N, factors)

In [7]:
def ShorNoQPU(N, coprime):
    # Classical replacement for the quantum part of Shor
    repeat_period_candidates = []
    work = 1
    precision_bits = math.ceil(math.log2(N))
    max_loops = pow(2, precision_bits)
    for iter in range(max_loops):
        work = (work * coprime) % N
        if work == 1: # found a repeat period
            repeat_period_candidates.append(iter + 1)
    return repeat_period_candidates

In [8]:
def ShorQPU(N, coprime, simulate=True, vis=False, opt_for_vis=False):
    # Quantum part of Shor's algorithm
    # For this implementation, the coprime must be 2.
    coprime = 2;
    return ShorQPU_WithModulo(N, coprime, simulate, vis, opt_for_vis)

    '''
    # For some numbers (like 15 and 21) the "mod" in a^xmod(N)
    # is not needed, because a^x wraps neatly around. This makes the
    # code simpler, and much easier to follow.
    if N == 15 or N == 21:
        return ShorQPU_WithoutModulo(N, coprime, simulate, opt_for_vis)
    else:
        return ShorQPU_WithModulo(N, coprime, simulate, vis, opt_for_vis)
    '''

In [9]:
def RollLeft(qc, work, num_shifts, control):
    # Ensure we're not out of range for work qubits
    for j in range(work.size - num_shifts-1,-1,-1):
        # Apply CSWAP between work[j] <-> work[j + num_shifts]
        qc.append(CSwapGate(), [control, work[j], work[j + num_shifts]])
    return qc

In [10]:
# In case our QPU read returns a "signed" negative value,
# convert it to unsigned.
def read_unsigned(qc, qreg, creg, simulate=True):
    qc.measure(qreg, creg)

    if simulate:
        # Simulate the circuit using the Aer simulator
        simulator = AerSimulator()
        job = simulator.run(transpile(qc, simulator), shots=1)

    else:
        # Run the circuit on a IBM quantum computer
        service = QiskitRuntimeService()
        #backend = service.least_busy(operational=True, simulator=False, min_num_qubits=qc.num_qubits)
        backend = service.backend("ibm_brisbane")
        print(backend.name)
        sampler = Sampler(backend)
        job = sampler.run([transpile(qc, backend)], shots=1)

    result = job.result()
    counts = result.get_counts(qc)

    # Visualize the results
    #print("Measurement results: " + str(counts) + "\n")
    #fig = plot_histogram(counts)
    #fig.show()

    # Extract value of most likely measurement outcome
    value = int(max(counts, key=counts.get),2)

    return value & ((1 << qreg.size) - 1)

In [11]:
def estimate_num_spikes(spike, spike_range):
    if spike < spike_range / 2:
        spike = spike_range - spike
    best_error = 1.0
    e0 = 0
    e1 = 0
    e2 = 0
    actual = spike / spike_range
    candidates = []
    for denominator in range(1,spike):
        numerator = round(denominator * actual)
        estimated = numerator / denominator
        error = abs(estimated - actual)
        e0 = e1
        e1 = e2
        e2 = error
        # Look for a local minimum which beats our current best error
        if e1 <= best_error and e1 < e0 and e1 < e2:
            repeat_period = denominator - 1
            candidates.append(repeat_period)
            best_error = e1
    return candidates

In [12]:
# This is the short/simple version of ShorQPU() where we can perform a^x and
# don't need to be concerned with performing a quantum int modulus.

def ShorQPU_WithoutModulo(N, coprime, simulate=True, vis=False, opt_for_vis=False):

    # Define number of required qubits
    N_bits = math.ceil(math.log2(N))
    work_qubits = N_bits + 1
    precision_qubits = N_bits

    # Set up the QPU and the working registers
    work = QuantumRegister(work_qubits, name="work")
    precision = QuantumRegister(precision_qubits, name="precision")
    classical = ClassicalRegister(precision_qubits, name="precision_measure")
    qc = QuantumCircuit(work, precision, classical)

    # Initialization: set the working register to state |1>
    # and the precision register to state |0>
    # put the precision register into superposition using Hadamard gate
    qc.initialize(1, work)
    qc.initialize(0, precision)
    qc.h(precision)

    if opt_for_vis: qc.barrier()

    # Perform 2^x for all possible values of x in superposition
    for iter in range(precision_qubits):
        num_shifts = 1 << iter
        if num_shifts < work_qubits:
            qc = RollLeft(qc, work, num_shifts, precision[iter])
            if opt_for_vis: qc.barrier()

    # Quantum Fourier Transform on precision register
    qc.append(QFT(precision_qubits), precision)

    if opt_for_vis: qc.barrier()

    if vis: print(qc)

    read_result = read_unsigned(qc, precision, classical, simulate)
    print('QPU read result: '+str(read_result)+'\n')
    repeat_period_candidates = estimate_num_spikes(read_result, 1 << precision_qubits)

    print('Circuit depth: ', qc.depth())
    print('Number of qubits: ', qc.num_qubits)

    return repeat_period_candidates

In [13]:
def add_int(qc, qreg, N, conditions):
    reverse_to_subtract = False
    if N == 0:
        return
    elif N < 0:
        N = -N
        reverse_to_subtract = True
    ops = []
    add_val = int(N)
    condition_mask = (1 << len(qreg)) - 1

    add_val_mask = 1
    while add_val_mask <= add_val:
        cmask = condition_mask & ~(add_val_mask - 1)
        if add_val_mask & add_val:
            add_shift_mask = 1 << (len(qreg) - 1)
            while add_shift_mask >= add_val_mask:
                cmask &= ~add_shift_mask
                ops.append((add_shift_mask, cmask))
                add_shift_mask >>= 1
        condition_mask &= ~add_val_mask
        add_val_mask <<= 1
    if reverse_to_subtract:
        ops.reverse()
    for inst in ops:
        op_qubits = []
        for i in range(len(qreg)):
            if inst[1] & (1 << i):
                op_qubits.append(qreg[i])
        for i in range(len(qreg)):
            if inst[0] & (1 << i):
                op_qubits.append(qreg[i])
        qc = multi_cx(qc, op_qubits, conditions)

    return qc


def multi_cx(qc, qubits, conditions=None, do_cz=False):
    ## This will perform a CCCCCX with as many conditions as we want
    ## The last qubit in the list is the target.

    target = qubits[-1]
    conds = qubits[:-1]
    ops = []

    controll_qubits = []

    for i in range(len(conds)):
        controll_qubits.append(conds[i])
    if conditions != None:
        for i in range(len(conditions)):
            controll_qubits.append(conditions[i])

    if len(controll_qubits) == 0:
        qc.x(target)
    else:
        qc.mcx(controll_qubits, target)

    return qc

In [14]:
# This is the complicated version of ShorQPU() where we DO
# need to be concerned with performing a quantum int modulus.
# That's a complicated operation, and it also requires us to
# do the shifts one at a time.
def ShorQPU_WithModulo(N, coprime, simulate=True, vis=False, opt_for_vis=False):
    max_value = 1
    mod_engaged = False

    # Define number of required qubits
    N_bits = math.ceil(math.log2(N))
    work_qubits = N_bits + 1
    precision_qubits = N_bits
    scratch_qubits = 1

    # Set up the QPU and the working registers
    work = QuantumRegister(work_qubits, name="work")
    precision = QuantumRegister(precision_qubits, name="precision")
    scratch = QuantumRegister(scratch_qubits, name="scratch")
    classical = ClassicalRegister(precision_qubits, name="precision_measure")
    qc = QuantumCircuit(work, precision, scratch, classical)

    # Initialization: set the working register to state |1>
    # and the precision register to state |0>
    # put the precision register into superposition using Hadamard gate
    qc.initialize(1, work)
    qc.initialize(0, precision)
    qc.initialize(0, scratch)
    qc.h(precision)

    if opt_for_vis: qc.barrier()

    N_sign_bit = work[work_qubits - 1]

    for iter in range(precision_qubits):
        condition_qubit = precision[iter]  # Get the condition qubit

        shifts = 1 << iter
        for shift in range(shifts):
            qc = RollLeft(qc, work, 1, precision[iter]) # multiply by the coprime 2
            if opt_for_vis: qc.barrier()
            max_value <<= 1

            if max_value >= N:
                mod_engaged = True

            if mod_engaged:
                qc = add_int(qc, work, -N, [condition_qubit]) # Substract N from working register, causing this to go negative if work < N
                if opt_for_vis: qc.barrier()
                qc.ccx(N_sign_bit, condition_qubit, scratch) # Skim off the sign bit
                if opt_for_vis: qc.barrier()
                qc = add_int(qc, work, N, [scratch[0], condition_qubit]) # If we went negative, undo the subtraction by adding N
                if opt_for_vis: qc.barrier()
                qc.x(work[0])
                qc.ccx(work[0], condition_qubit, scratch) # If it's odd, then we wrapped, so clear the wrap bit
                qc.x(work[0])
                if opt_for_vis: qc.barrier()

    # Quantum Fourier Transform on precision register
    qc.append(QFT(precision_qubits), precision)

    if vis: print(qc)

    read_result = read_unsigned(qc, precision, classical, simulate)
    print('QPU read result: ', read_result)
    repeat_period_candidates = estimate_num_spikes(read_result, 1 << precision_qubits)

    print('Circuit depth: ', qc.depth())
    print('Number of qubits: ', qc.num_qubits)

    return repeat_period_candidates

# Execution

**Here are some values of N to try:**

15, 21, 33, 35, 39, 51, 55, 69, 77, 85, 87, 91, 93, 95, 111, 115, 117,
119, 123, 133, 155, 187, 203, 221, 247, 259, 287, 341, 451

**Larger numbers require more bits of precision.**

N = 15    precision_bits >= 4

N = 21    precision_bits >= 5

N = 35    precision_bits >= 6

N = 123   precision_bits >= 7

N = 341   precision_bits >= 8  time: about 6 seconds

N = 451   precision_bits >= 9  time: about 23 seconds

In [ ]:
def shor_sample(classical=False, simulate=True, vis=False, opt_for_vis=False):
    N = 33             # The number we're factoring
    coprime = 2        # Needs to be 2 in this implementation

    max_iteration = 10
    success = 0
    calculation_time = []

    for i in range(max_iteration):
        start_time = time.time()
        result = Shor(N, coprime, classical, simulate, vis, opt_for_vis)
        end_time = time.time()
        if result != None:
            success += 1
            calculation_time.append(end_time-start_time)
            print('Iteration '+str(i+1)+': Successful'+'. Calculation time: '+str("{:.2f}".format(end_time-start_time))+'s'+'. Success: '+str(success)+'/'+str(max_iteration))

            if success == 1: # Store solution
                solution = str(N)+'='+str(result[0])+'*'+str(result[1])
        else:
            print('Iteration '+str(i+1)+': Not Successful'+'. Success: '+str(success)+'/'+str(max_iteration))
        print('-----------------------------------------------------------------------------------\n')

    if success == 0:
        print('Failure: No non-trivial factors were found.\n')
    else:
        success_rate = success/max_iteration
        average_time = sum(calculation_time)/len(calculation_time)
        print('Solution: '+solution+'\nAccuracy: '+str(success_rate*100)+'%'+'\nAverage calculation time: '+str("{:.2f}".format(average_time))+'s')

# 'classical=False' executes quantum part of algorithm
# 'classical=True' executes classical alternative of quantum part for comparison
# 'simulate=True' obtains the result by running the circuit on a simulator
# 'simulate=False' obtains the result by running the circuit on a IBM quantum computer
# 'vis=True' shows quantum circuit
# 'vis=False' does not show quantum circuit
# 'opt_for_vis=True' adds barriers to the circuit to improve readability
# 'opt_for_vis=False' removes barriers in the circuit to improve performance
shor_sample(classical=False, simulate=False, vis=False, opt_for_vis=False)

ibm_brisbane


Solution: 341=31*11
Circuit depth:  35763
Number of qubits:  20
Accuracy: 90.0%
Average calculation time: 151.11s